addons/netfox/properties/property-store-snapshot.gd

```diff
@@ -0,0 +1,83 @@
+extends RefCounted
+class_name PropertyStoreSnapshot
+
+# Typed as Dictionary[String, Property]
+var _snapshot : Dictionary = {}
+
+func size() -> int:
+	return _snapshot.size()
+
+func get_paths() -> Array[String]:
+	var array: Array[String] = []
+	for key in _snapshot.keys():
+		array.push_back(key as String)
+	return array
+
+func get_property(key: String) -> Property:
+	return _snapshot[key]
+
+func get_properties() -> Array[Property]:
+	var array: Array[Property] = []
+	for property in _snapshot.values():
+		array.push_back(property as Property)
+	return array
+
+func set_property(key: String, value: Property):
+	_snapshot[key] = value;
+
+func remove_property(path: String) -> bool:
+	return _snapshot.erase(path)
+
+func get_snapshot() -> Dictionary:
+	return _snapshot
+
+func has_key(key: String) -> bool:
+	return _snapshot.has(key)
+
+func serialize() -> Dictionary:
+	var serialized: Dictionary = {}
+	for property in _snapshot:
+		serialized[property] = _snapshot[property].serialize()
+	return serialized
+
+static func deserialize(data: Dictionary) -> PropertyStoreSnapshot:
+	var deserialized := PropertyStoreSnapshot.new()
+	for property in data:
+		deserialized.set_property(property, Property.deserialize(data[property]))
+	return deserialized
+
+func is_empty() -> bool:
+	return _snapshot.is_empty()
+
+# b.get_diffs_from(a)
+# Where b is the newer form of a, and we only want what changed in b
+
+func get_diffs_from(reference: PropertyStoreSnapshot) -> PropertyStoreSnapshot:
+	var diffs := PropertyStoreSnapshot.new()
+	
+	for key in _snapshot.keys():
+		if not reference.has_key(key):
+			diffs.set_property(key, _snapshot[key])
+		elif reference.get_property(key) != _snapshot[key]:
```

Comment:
I think this check doesn't work, since instead of comparing the property values, we're comparing `Property` instances. These are compared by reference, so this check will always be true.